File: turn-orchestrator/src/subscriber.rs
//! Subscriber on `turn::step_requested`. One trigger event = one state
//! transition. After running the transition the subscriber re-publishes
//! the topic if the record is not terminal.

use std::sync::Arc;

use iii_sdk::{IIIError, RegisterFunctionMessage, Value, III};
use serde_json::json;

use crate::config::TurnOrchestratorConfig;
use crate::persistence;
use crate::run_start::publish_step;
use crate::transitions;

pub const FUNCTION_ID: &str = "turn::step";

pub async fn execute(
    iii: III,
    cfg: Arc<TurnOrchestratorConfig>,
    payload: Value,
) -> Result<Value, IIIError> {
    let session_id = extract_session_id(&payload).ok_or_else(|| {
        IIIError::Handler("turn::step_requested payload missing session_id".into())
    })?;

    let mut record = match persistence::load_record(&iii, &session_id).await {
        Some(r) => r,
        None => {
            tracing::warn!(%session_id, "turn::step_requested for unknown session");
            return Ok(json!({ "ok": false, "reason": "unknown_session" }));
        }
    };

    if record.is_terminal() {
        return Ok(json!({ "ok": true, "terminal": true }));
    }

    let from_state = record.state;
    transitions::step(&iii, &cfg, &mut record)
        .await
        .map_err(|e| {
            IIIError::Handler(format!(
                "transition from {} failed: {e}",
                from_state.as_str()
            ))
        })?;
    persistence::save_record(&iii, &record).await;

    if !record.is_terminal() {
        publish_step(&iii, &session_id).await;
    }
    Ok(json!({
        "ok": true,
        "from_state": from_state.as_str(),
        "to_state": record.state.as_str(),
    }))
}

pub fn register(iii: &III, cfg: &Arc<TurnOrchestratorConfig>) {
    let iii_for_handler = iii.clone();
    let cfg_for_handler = Arc::clone(cfg);
    iii.register_function((
        RegisterFunctionMessage::with_id(FUNCTION_ID.to_string()).with_description(
            "Run one durable state machine transition for a session.".to_string(),
        ),
        move |payload: Value| {
            let iii = iii_for_handler.clone();
            let cfg = Arc::clone(&cfg_for_handler);
            async move { execute(iii, cfg, payload).await }
        },
    ));
}

fn extract_session_id(payload: &Value) -> Option<String> {
    // `iii::durable::publish` delivers the `data` object here. Accept optional extra nesting
    // (`payload` / `data`) for compatibility with older in-memory pub/sub shapes.
    let inner = payload
        .get("payload")
        .or_else(|| payload.get("data"))
        .unwrap_or(payload);
    inner
        .get("session_id")
        .and_then(Value::as_str)
        .map(str::to_string)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn subscriber_register_accepts_config_arc() {
        // Compile-time pin: register() must take an Arc<TurnOrchestratorConfig>.
        // This guards against silently dropping config plumbing.
        fn _assert_signature(
            iii: &iii_sdk::III,
            cfg: &std::sync::Arc<crate::config::TurnOrchestratorConfig>,
        ) {
            super::register(iii, cfg);
        }
    }

    #[test]
    fn extract_session_id_from_nested_payload() {
        let p = json!({ "payload": { "session_id": "s1" }});
        assert_eq!(extract_session_id(&p).as_deref(), Some("s1"));
    }

    #[test]
    fn extract_session_id_from_top_level() {
        let p = json!({ "session_id": "s2" });
        assert_eq!(extract_session_id(&p).as_deref(), Some("s2"));
    }

    #[test]
    fn extract_session_id_returns_none_when_missing() {
        assert!(extract_session_id(&json!({})).is_none());
    }
}

File: harness-node/src/turn-orchestrator/subscriber.ts
/**
 * `turn::step` — one FSM transition for a session.
 *
 * **Incoming**: flat `{ session_id }` from durable subscriber (`turn::step_requested`),
 * direct `iii.trigger('turn::step', …)`, and integration tests — same shape.
 * Producers publish via `iii::durable::publish` with `{ topic, data: { session_id } }`;
 * the engine enqueues `data` only.
 *
 * **Outgoing**: `StepResult` with pre/post `TurnState`; throws on missing session
 * (invariant) or transition failure. `turn::should_step` soft-filters unknown/terminal
 * sessions before the durable subscriber invokes `turn::step`.
 */

import type { StateGetInput } from 'iii-sdk/state';
import { z } from 'zod';
import type { ISdk } from '../runtime/iii.js';
import { logger } from '../runtime/otel.js';
import type { TurnOrchestratorConfig } from './config.js';
import * as persistence from './persistence.js';
import { isTerminal, turnStateKey, type TurnState, type TurnStateRecord } from './state.js';
import { step } from './transitions.js';

export const StepPayloadSchema = z.object({
  session_id: z.string().min(1),
});

export type StepPayload = z.infer<typeof StepPayloadSchema>;

export type StepResult = { ok: true; from_state: TurnState; to_state: TurnState };

export async function shouldStep(iii: ISdk, payload: unknown): Promise<boolean> {
  const parsed = StepPayloadSchema.safeParse(payload);
  if (!parsed.success) return false;
  const rec = await persistence.loadRecord(iii, parsed.data.session_id);
  if (!rec) {
    logger.warn('turn::step for unknown session', { session_id: parsed.data.session_id });
    return false;
  }
  return !isTerminal(rec);
}

export async function execute(
  iii: ISdk,
  cfg: TurnOrchestratorConfig,
  payload: StepPayload,
): Promise<StepResult> {
  const { session_id } = payload;
  const rec = await iii.trigger<StateGetInput, TurnStateRecord>({
    function_id: 'state::get',
    payload: { scope: 'agent', key: turnStateKey(session_id) },
  });
  const from_state = rec.state;
  try {
    await step(iii, cfg, rec);
  } catch (err) {
    throw new Error(`transition from ${from_state} failed: ${String(err)}`);
  }
  await persistence.saveRecord(iii, rec);
  return { ok: true, from_state, to_state: rec.state };
}

export function register(iii: ISdk, cfg: TurnOrchestratorConfig): void {
  iii.registerFunction('turn::should_step', async (payload: unknown) => shouldStep(iii, payload), {
    description:
      'Condition: durable turn::step_requested payload has a known, non-terminal session.',
  });

  iii.registerFunction(
    'turn::step',
    async (payload: unknown) => execute(iii, cfg, StepPayloadSchema.parse(payload)),
    {
      description: 'Run one durable state machine transition for a session.',
    },
  );
  iii.registerTrigger({
    type: 'durable:subscriber',
    function_id: 'turn::step',
    config: {
      topic: 'turn::step_requested',
      condition_function_id: 'turn::should_step',
    },
  });
}
